src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs

```diff
@@ -506,7 +507,8 @@ public bool SdamEquals(ServerDescription other)
                 _replicaSetConfig?.Version == other._replicaSetConfig?.Version && // setVersion
                 object.Equals(_electionId, other._electionId) &&
                 EndPointHelper.Equals(_replicaSetConfig?.Primary, other._replicaSetConfig?.Primary) && // primary
-                _logicalSessionTimeout == other._logicalSessionTimeout;
+                _logicalSessionTimeout == other._logicalSessionTimeout &&
+                ExceptionHelper.Equals(_heartbeatException, other._heartbeatException);
```

Comment:
Replace with:

    SdamEquals(_heartbeatException, other._heartbeatException);

where `SdamEquals` is a local helper method in this class:

    private bool SdamEquals(Exception x, Exception y)
    {
        if (x == null && y == null)
        {
            return true;
        }
        if (x == null || y == null)
        {
            return false;
        }

        return
            x.GetType() == y.GetType() &&
            string.Equals(x.Message, y.Message) &&
            string.Equals(x.StackTrace, y.StackTrace) &&
            SdamEquals(x.InnerException, y.InnerException);
    }

Reply:
Note: I recommend a regular (private) method instead of a local function so that you can test it.